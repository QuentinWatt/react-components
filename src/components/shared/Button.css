@tailwind components;
@tailwind utilities;

@layer components {
  .button {
    @apply bg-blue-500 hover:bg-blue-600 rounded flex items-center justify-center px-3 py-2 text-white min-w-32;
  }
}
