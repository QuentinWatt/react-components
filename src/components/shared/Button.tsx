import { MouseEventHandler } from "react";
import Loader from "./Loader";
import "./Button.css";

interface Props {
  children: React.ReactNode;
  className?: string;
  loading?: boolean;
  onClick?: MouseEventHandler<HTMLButtonElement>;
  type?: "button" | "reset" | "submit";
}

const Button: React.FC<Props> = ({
  children,
  className = "",
  onClick = () => {},
  type = "button",
  loading = false,
}) => {
  return (
    <button className={`button ${className}`} type={type} onClick={onClick}>
      {loading ? <Loader /> : children}
    </button>
  );
};

export default Button;
