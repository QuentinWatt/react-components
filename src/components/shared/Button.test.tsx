import "@testing-library/jest-dom";
import { render, screen, fireEvent } from "@testing-library/react";
import Button from "./Button";

jest.mock("./Loader", () => () => <div data-testid="loader-mock">loader</div>);
jest.mock("./Loader.css", () => () => {});

describe("Shared button component", () => {
  it("renders with children, and props as button text", () => {
    const onClickMock = jest.fn();
    const { asFragment } = render(
      <Button onClick={onClickMock} type="submit" className="test-class">
        Press me!
      </Button>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders with children as button text", () => {
    render(<Button>Press me!</Button>);
    const element = screen.getByText(/Press me!/i);
    expect(element).toBeInTheDocument();
  });

  it("renders a loader if the loading prop is passed", () => {
    render(<Button loading={true}>Press me!</Button>);
    const loaderElement = screen.getByTestId("loader-mock");
    expect(loaderElement).toBeTruthy();

    expect(screen.queryByText(/Press me!/i)).toBeNull();
  });

  it("defaults to type button", () => {
    render(<Button>Press me!</Button>);
    const button = screen.getByText(/Press me/i);
    expect(button).toHaveAttribute("type", "button");
  });

  it("changes the button type with a prop", () => {
    render(<Button type="submit">Press me!</Button>);
    const button = screen.getByText(/Press me/i);
    expect(button).toHaveAttribute("type", "submit");
  });

  test("renders with custom className", () => {
    const { getByText } = render(
      <Button className="custom-class">Click Me</Button>
    );
    const button = getByText(/Click Me/i);
    expect(button).toHaveClass("custom-class");
  });

  it("calls onClick handler when clicked", () => {
    const onClickMock = jest.fn();
    const { getByText } = render(
      <Button onClick={onClickMock}>Press me!</Button>
    );

    const button = getByText(/Press me!/i);
    fireEvent.click(button);
    expect(onClickMock).toHaveBeenCalled();
  });
});
