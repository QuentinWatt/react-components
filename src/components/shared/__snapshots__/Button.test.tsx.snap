// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shared button component renders with children, and props as button text 1`] = `
<DocumentFragment>
  <button
    class="button test-class"
    type="submit"
  >
    Press me!
  </button>
</DocumentFragment>
`;
